wip/beam-spec.md

```diff
@@ -0,0 +1,437 @@
+execution of the function that was preempted by the interrupt handler.)
+
+- BEAM's interrupt handling capabilities can be summarized as: prioritization of
+  interrupts, individual interrupt masking and global interrupt masking.
+
+- The BEAM has 8 *interrupts* whose *handlers* are named: `INTERRUPT0`,
+  `INTERRUPT1`, all the way to `INTERRUPT7`.
+
+- The signature of all interrupt handlers must be `[unsafe] fn()`. That is these
+  handlers don't take arguments nor can they return values.
+
+- If not defined in the program, interrupt handlers default to:
+
+``` rust
+#[no_mangle]
+unsafe fn INTERRUPT_() {
+    loop {}
+}
+```
+
+- When the BEAM starts, all interrupts are *disabled*, *unmasked* and are
+  configured with an initial *priority* of `1`; and the *running priority* of
+  the system is set to `0`.
+
+- Interrupts can be assigned any *priority* in the range `1..=255`.
+
+- At any point in time an interrupt signal may arrive and change the state of an
+  interrupt to *pending*. That is interrupt signals occur asynchronously to the
+  execution of the program.
+
+- If interrupts are *enabled*, setting the state of an *unmasked* interrupt to
+  *pending* results in its handler preempting the current function *if and only
+  if* the *priority* of the interrupt is *higher* than the current *running
+  priority*, otherwise the interrupt will be kept in the *pending* state until
+  the *running priority* is low enough to allow preemption.
+
+- The *pending* state of an interrupt gets cleared when its handler starts
+  executing.
+
+- Once an interrupt handler starts executing the *running priority* is
+  automatically set to the *priority* of the interrupt being serviced. When the
+  interrupt handler returns the running priority is restored to its previous
+  (before preemption) value.
+
+- While interrupts are *disabled* preemption can not occur. Interrupts in the
+  pending state will remain in that state until interrupts are re-enabled.
+
+- Setting the state of a *masked* interrupt to *pending* will never result in
+  preemption, no matter the priority of the interrupt. Such interrupt will only
+  be serviced after it has been unmasked.
+
+- The behavior of setting the *running priority* within the context of an
+  interrupt handler to a value *lower* than the *priority* of that interrupt
+  results in UNSPECIFIED behavior.
+
+- Changing the *priority* of an interrupt from within its handler results in
+  UNSPECIFIED behavior.
+
+- Interrupt handling can be summarized as follows:
+
+``` rust
+// pseudo-code
+if interrupts_are_enabled {
+    if interrupt_N_is_pending {
+        if !interrupt_N_is_masked {
+            if interrupt_N_priority > running_priority {
+                atomic {
+                    interrupt_N_is_pending = false;
+                    running_priority = interrupt_N_priority;
+                    start_INTERRUPTN();
+                }
+            }
+        }
+    }
+}
+```
+
```

Comment:
From the text, it is obvious that preemption can occur in the middle of a function, but I was wondering, at which precise granularity can preemption occur ? Or maybe in different words, what are the "atomic" operations of the abstract machine, during which preemption does not occur ? 

For example. consider a single instruction like `add` in an architecture in which it returns a result in one register, and sets a flag in another register indicating whether the add overflowed. Now consider two arguments for which the `add` overflows. Could the execution of this instruction be preempted after setting the flag register but before writing the result value? That would allow an interrupt handler to remove the overflow flag, and would prevent the interrupted code from knowing that an overflow happened. 

This should not be able to happen, that is, such an `add` operation that modifies multiple registers would be non-preemptable in the middle of its execution. Or in other words, preemption can only happen across these non-preemptable operations.

I don't really know what's the best way to declare / specify at which granularity can pre-emption happen, but this should be specified somewhere.

Reply:
For example, in a different less-useful abstract machine, preemption could only be allowed across, for example, function calls (e.g. in BEAM, this would be like setting the priority to 255 when entering every function, and setting it to 0 right before calling a function, such that the only window in which this can happen, is right before entering the next function).

Reply:
>consider a single instruction like add in an architecture in which it returns a result in one register, and sets a flag in another register indicating whether the add overflowed. Now consider two arguments for which the add overflows. Could the execution of this instruction be preempted after setting the flag register but before writing the result value?

I'm not sure about specifics of embedded hardware, but this seems to be pretty unlikely behavior.

CPU ISAs are normally specified in terms of instructions following each other in-order and changing observable architectural state like registers atomically (unless explicitly said otherwise).

Even if CPU works is out-of-order fashion internally, instructions as defined by ISA are still retired in-order and maintain architectural state at retirement.

That's why taking of interrupts/exceptions is also tied to instruction retirement, and the first instruction of exception handler will see consistent state left by the previous retired instruction (this is called "precise interrupts").

TLDR: the granularity is macro instructions as defined by respective CPU's ISA.

Reply:
@petrochenkov Yeah, I guess something like MIR, where `add_with_overflow` is an actual intrinsic, would be more interesting to consider than an actual ISA then. MIR intrinsics are going to get lowered into multiple CPU instruction, and preemption can happen across any of them.

Reply:
This is getting into the interaction of hardware-level interrupts and language abstract machines... that's a *hard* problem.^^  This is why C says that in an interrupt, the content of all memory is unspecified except for volatile and atomic stuff. We likely need something similar.